plugins/notifications-backend/src/service/handlers.ts

```diff
@@ -0,0 +1,111 @@
+import { Knex } from "knex";
+import { CreateNotificationRequest, Notification, NotificationsFilter } from "./types";
+import { CatalogClient } from "@backstage/catalog-client";
+import { MessagesInsert } from "./db";
+
+// createNotification
+// returns string id of created notification
+export async function createNotification ( dbClient: Knex<any, any>, catalogClient: CatalogClient
+    , req: CreateNotificationRequest
+    ): Promise<{msgid: string}> {
+    if ( req.targetGroups ) {
+        req.targetGroups.forEach(async group => {
+            const groupRef = await catalogClient.getEntityByRef("group:"+group);
+            if ( !groupRef ) {
+                throw new Error(`group '${group}' does not exist`);
+            }
+        });
+    }
+
+    if ( req.targetUsers ) {
+        req.targetUsers.forEach(async user => {
+            const groupRef = await catalogClient.getEntityByRef("user:"+user);
+            if ( !groupRef ) {
+                throw new Error(`user '${user}' does not exist`);
+            }
+        });
+    }
+
+    const row: MessagesInsert = {
+        origin: req.origin,
+        title: req.title,
+        message: req.message,
+        topic: req.topic
+    };
+
+    let msgid: string = '';
+
+    await dbClient('messages')
+      .insert(row)
+      .returning('id')
+      .then(id => {
+        msgid = id as unknown as string;
+      });
+
+    return {msgid: msgid};
+}
+
+// getNotifications
+export async function getNotifications( dbClient: Knex<any, any>
+    , filter: NotificationsFilter
+    , pageSize: number
+    , pageNumber: number
+    ): Promise<Notification[]> {
+
+    if ( pageSize < 0 || pageNumber < 0  ||
+         pageSize == 0 && pageNumber > 0 ||
+         pageSize > 0 && pageNumber == 0 ) {
+        throw new Error('pageSize and pageNumber must both be either 0 or greater than 0');
+    }
+
+    let query = dbClient('messages').select('*');
+
+    addFilter(filter, query);
```

Comment:
nit: Feel free to ignore
If we flip the params to `addFilter(query, filter)`, it will be easier to extend the list in the future.

Reply:
fixed